src/core/scripthost.cpp

```diff
@@ -368,8 +363,48 @@ void ScriptHost::resetWatches()
 
 bool ScriptHost::autoTrack()
 {
-    // This is called every frame to run auto-tracking
-    // returns true if auto-tracking changed stuff, false otherwise
-    if (_autoTracker) return _autoTracker->doStuff();
+    if (_autoTracker && _autoTracker->doStuff()) {
+        // autotracker changed the cache
+
+        // run callbacks
+        runMemoryWatchCallbacks();
+
+        return true;
+    }
+
+    // autotracker didn't change anything
     return false;
 }
+
+void ScriptHost::runMemoryWatchCallbacks()
+{
+    // we need to run callbacks because the autotracker changed some cache
+    // but really we only need to run the ones that are marked dirty
+    for (size_t i = 0; i < _memoryWatches.size(); i++) {
+        // NOTE: since watches can change in a callback, we use vector
+        auto& w = _memoryWatches[i];
+        auto name = w.name;
+        auto data = w.data;
+
+        // skip this watch if it's not dirty
+        if (!w.dirty)
+            continue;
+
+        // run memory watch callback
+        lua_rawgeti(_L, LUA_REGISTRYINDEX, w.callback);
+        _autoTracker->Lua_Push(_L); // arg1: autotracker ("segment")
+        if (lua_pcall(_L, 1, 0, 0)) {
+            printf("Error calling Memory Watch Callback for %s: %s\n",
+                w.name.c_str(), lua_tostring(_L, -1));
+            lua_pop(_L, 1);
+        }
+
+        // TODO - clear dirty flag
```

Comment:
can we maybe make this depend on the backend in use? I think i wouldn't want to change the behaviour of snes since people should write packs in a way where the current code works.

Reply:
If my understanding is correct, this change to ScriptHost is necessary if we want it to behave more like Emotracker, which would reduce the difference in pack behavior between the two applications.

See this snippet from Codetracker which uses the return value to indicate that the callback should continue to trigger until the player is in game. The current code ignores the return value, so it won't necessarily trigger the callback again after the player begins the game. The underlying memory did not change, but the callback hasn't actually read it yet. This is especially important because this condition is even more likely to occur in Poptracker since our implementation of ReadU8 does not bypass the cache.

```
function isInGame()
    local module = AutoTracker:ReadU8(0x7e0010, 0)
    return module > 0x05 and module < 0x55 and module ~= 0x14
end

    <callback snip>
    if not isInGame() then
        return false
    end
```

Reply:
Oh, that's interesting. So we should definitely implement the return value handling.... and document it.
While the documentation does not look especially good, it is complete enough to have helped a ton of people. It has become an important part of Pop.